### PR TITLE
Service groups (like kubernetes pods)

### PR DESCRIPTION
**Pull request by ravenac95**
_This is a pretty massive change. I need to add some additional tests, but this is something pretty critical for our use-case and I imagine other company's use cases._

# Service Groups For Convox

## Summary of Changes

* Introduces Service Groups
* Introduces Secure Environment

### Service Groups

This PR introduces service groups for convox. A service group allows you to create something akin to a pod in kubernetes. It allows you to group sets of docker containers that will be deployed together in a single ECS task. Imagine you have a simple `docker-compose.yml` file that looks like this:

```
version: "2"

services:
  one:
    build: .
    ports:
      - 4444:4444
    labels:
      - convox.port.4444.protocol=http
      - convox.health.path=/status
      - convox.health.port=4444

  two:
    build: .
    ports:
      - 80:80
    labels:
      - convox.balancer=false
  
  three: 
    build: .
    ports:
      - 80:80
    labels:
      - convox.balancer=false

  four:
    build: .
    ports:
      - 80:80
    labels:
      - convox.balancer=false
```

In order to group containers `two` and `three` to container `one`, you would add a `convox.group=[group_name]` label to all of the containers, like so:

```
version: "2"

services:
  one:
    build: .
    ports:
      - 4444:4444
    links:
      - two
      - three
    labels:
      - convox.group=main
      - convox.port.4444.protocol=http
      - convox.health.path=/status
      - convox.health.port=4444

  two:
    build: .
    expose:
      - 80
    labels:
      - convox.group=main
      - convox.balancer=false
  
  three: 
    build: .
    expose:
      - 80
    labels:
      - convox.group=main
      - convox.balancer=false

  four:
    build: .
    ports:
      - 80:80
    labels:
      - convox.balancer=false
```

Now `one`, `two`, and `three` are part of the `main` group. Notice in the second `docker-compose.yml` file that `two` and `three` no longer need to bind ports onto the host, they can simply expose ports and be linked to container `one`.

##### Using `convox ps`

The PIDs for containers is now using a two part string split by a `-`. This allows for us to choose a specific container within a specific ECS task. The first part of the string references the taskArn and the second part references the containerArn.

The output of the command looks like this:

```
$ convox ps
ID                         GROUP           NAME   RELEASE      STARTED    COMMAND
8882440bb4c7-75e2d9874eb7  secondarygroup  four   RBPHEISKCSO  1 day ago  node index.js
e84fa1cf951d-6a64c3abdcf6  maingroup       one    RBPHEISKCSO  1 day ago  node index.js
e84fa1cf951d-d84d9ac39395  maingroup       three  RBPHEISKCSO  1 day ago  node index.js
e84fa1cf951d-0dad943e9420  maingroup       two    RBPHEISKCSO  1 day ago  node index.js
```

##### Using `convox exec`

Due to the change in PID structure. In order to use convox exec. You must pass in the full PID like `e84fa1cf951d-0dad943e9420` above to use the `two` container in the `maingroup`

### Secure Environment

This PR also introduces a completely new idea to convox that I call the "Secure Environment" option for containers. One of the problems with service groups is that if you include many large environment variables for your convox apps all of those environment variables get listed in the ECS TaskDefinition. This causes a problem if your environment variables are sufficiently large (maybe you're putting things like JSON configurations or Certs) as ECS does not allow TaskDefinitions to exceed a size of 32K. The introduction of the Secure Environment kills two birds with one stone:

1. It circumvents the 32K limit on ECS TaskDefinitions (_Obviously the limit isn't completely defeated, but now it'd take a fairly large task definition to make that possible. Without this it's pretty trivial to do it with just a few grouped containers_)
2. It ensures that the environment will NOT appear on AWS's ECS console (required for my use case). This plus #2052 and #2097 will be the holy grail of a secure environment.

A secure environment can be specified for a container in the `docker-compose.yml` file by adding `convox.secure-env=true` to the labels, like so:

```
version: "2"

services:
  secure:
    build: .
    ports:
      - 4444:4444
    labels:
      - convox.secure-env=true
```

Once the secure environment is enabled, 3 environment variables will be passed into the container at runtime. These variables are: 

* `SECURE_ENVIRONMENT_URL` - The url to the s3 file that contains the environment (encrypted with KMS)
* `SECURE_ENVIRONMENT_TYPE` - The type of the file that is encrypted in s3. This is reserved for the future. For now it only returns `envfile`. _We may want to use some actual mimetype (or create one I guess)
* `SECURE_ENVIRONMENT_KEY` - The arn to the KMS key. Used for decryption of the environment file referenced by  `SECURE_ENVIRONMENT_URL`

The intent is that the container will download and decrypt the secure environment upon start up. The advantage for my team is that the environment can still be updated by the convox CLI and the convox console GUI but it will not be available to AWS's console. 

_I will update this PR soon with some instructions on how it's used and also add a link to a base container that can load the secure environment._

## TODOS

- [x] Add tests for `./manifests`
- [ ] Update the documentation
- [ ] Add github link to a repository for a container using the secure environment

## Bonus!

In order to develop this, I actually created a pretty useful testing tool that allowed me to deploy a completely custom convox deployment. Right now convox is pretty tied to some internal release process. This tool will alleviate that making it easier for others to contribute. I will also be making a PR for that tool as it was designed to integrate tightly with convox.

### CODE REVIEW

**Comment by ravenac95** [comment]
Oh apparently I'm out of date with the base branch. I'll make sure to update it.

File: provider/aws/processes_test.go [hunk 1]
```diff
@@ -1273,11 +1307,17 @@ var cycleProcessDockerInspect = awsutil.Cycle{
 	Response: awsutil.Response{
 		StatusCode: 200,
 		Body: `{
+			"Id": "8dfafdbc3a40",
+			"ID": "8dfafdbc3a40",
```
Comment (ravenac95) [inline provider/aws/processes_test.go]: Wasn't sure which id value to use here. I think `ID` is correct

File: api/awsutil/awsutil.go [hunk 2]
```diff
@@ -92,6 +92,7 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		io.WriteString(w, cycle.Response.Body)
 	} else {
 		fmt.Println("Request does not match next cycle.")
+		fmt.Printf("Remaining cycles: %d\n", len(h.cycles))
```
Comment (ravenac95) [inline api/awsutil/awsutil.go]: The addition of this line made fixing the tests for the aws provider worlds easier.

**Comment by ravenac95** [comment]
Somehow the build started failing on CI when i made more changes. It works locally. I'll rebase to master and push it all back up.

**Comment by MiguelMoll** [comment]
[Service groups (like kubernetes pods)](https://trello.com/c/Wspy8W2W/2235-service-groups-like-kubernetes-pods)

**Comment by nzoschke** [comment]
Wow, thanks for the patch. Going to take a look at this now and get back to you with feedback.

**Comment by nzoschke** [comment]
My first impressions of the features here are positive. I get the use cases for the features, and it looks like a strong start for the implementations.

The main concern is general maintenance. Will these new concepts get used and justify the additional surface area?

Some things I'd want to do to go forward:

* Split out Service Groups and Secure Env into separate PRs
* Review the changes to the `convox ps` API for task, container and group. Are there compatibility issues?
* Review the use labels. This pattern is becoming a default implementation strategy for tons of things and doesn't feel great.
* Add additional tests
* QA the features

@ravenac95 I reached out over email, I hope we can discuss all this over higher bandwidth soon.

**Comment by ravenac95** [comment]
@nzoschke I'm totally fine separating the Secure Env from the service group, but the one thing I worry about is around the size limits of ECS tasks which is why I coupled the deployment of the two features. That being said, if it's simply for PR simplicity I'm 100% for it. I would say however, that the two features should roll out at the same time and I'm not sure what your process would be for managing blocking features.

**Comment by MiguelMoll** [comment]
Excellent PR but going to close it for now as it's a bit old. We can use the code going forward when ready to implement a similar strategy. Thanks!

**Pull request closed without merging**